src/gainmap.c

```diff
@@ -49,4 +53,238 @@ avifBool avifGainMapMetadataFractionsToDouble(avifGainMapMetadataDouble * dst, c
+                const float gainMapLog = lerp(gainMapMinLog[c], gainMapMaxLog[c], powf(gainMapValue, (float)metadata.gainMapGamma[c]));
+                const float toneMappedLinear = (baseLinear + (float)offsetBase[c]) * expf(gainMapLog * weight) - (float)offsetOther[c];
+
+                if (toneMappedLinear > rgbMaxLinear) {
+                    rgbMaxLinear = toneMappedLinear;
+                }
+                if (toneMappedLinear > pixelRgbMaxLinear) {
+                    pixelRgbMaxLinear = toneMappedLinear;
+                }
+
+                const float toneMappedGamma = linearToGamma(toneMappedLinear);
+                toneMappedPixelRGBA[c] = AVIF_CLAMP(toneMappedGamma, 0.0f, 1.0f);
+            }
+            toneMappedPixelRGBA[3] = basePixelRGBA[3]; // Alpha is unaffected by tone mapping.
+            rgbSumLinear += pixelRgbMaxLinear;
+            avifSetRGBAPixel(toneMappedImage, i, j, &toneMappedPixelRGBState, toneMappedPixelRGBA);
+        }
+    }
+    if (clli != NULL) {
+        // For exact CLLI value definitions, see ISO/IEC 23008-2 section D.3.35
+        // at https://standards.iso.org/ittf/PubliclyAvailableStandards/index.html
+
+        // Convert extended SDR (where 1.0 is SDR white) to nits.
+        clli->maxCLL = (uint16_t)AVIF_CLAMP(avifRoundf(rgbMaxLinear * SDR_WHITE_NITS), 0.0f, (float)INT16_MAX);
+        const float rgbAverageLinear = rgbSumLinear / (width * height);
+        clli->maxPALL = (uint16_t)AVIF_CLAMP(avifRoundf(rgbAverageLinear * SDR_WHITE_NITS), 0, (float)INT16_MAX);
+    }
+
+cleanup:
+    avifRGBImageFreePixels(&rgbGainMap);
+    if (rescaledGainMap != NULL)
+        avifImageDestroy(rescaledGainMap);
+
+    return res;
+}
+
+avifResult avifImageApplyGainMap(const avifImage * baseImage,
+                                 const avifGainMap * gainMap,
+                                 float hdrCapacity,
+                                 avifTransferCharacteristics outputTransferCharacteristics,
+                                 avifRGBImage * toneMappedImage,
+                                 avifContentLightLevelInformationBox * clli,
+                                 avifDiagnostics * diag)
+{
+    avifRGBImage baseImageRgb;
+    avifRGBImageSetDefaults(&baseImageRgb, baseImage);
+    AVIF_CHECKRES(avifRGBImageAllocatePixels(&baseImageRgb));
+    avifResult res = avifImageYUVToRGB(baseImage, &baseImageRgb);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
+
+    res = avifImageApplyGainMapRGB(&baseImageRgb,
+                                   baseImage->transferCharacteristics,
+                                   gainMap,
+                                   hdrCapacity,
+                                   outputTransferCharacteristics,
+                                   toneMappedImage,
+                                   clli,
+                                   diag);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
```

Comment:
This if statement can be omitted. (Given the definition of this function, we are unlikely to add new code here.)